File: 1512_good_pairs/good_pairs.js
// JS (Hashmap): https://leetcode.com/problems/number-of-good-pairs/submissions/1188124896/
// TS (Iteration/brute force): https://leetcode.com/problems/number-of-good-pairs/submissions/1177868936/
function numberOfGoodPairs(nums) {
    let count = 0;

    for(let i = 0; i < nums.length; i++) {
        for(let j = j + i; j < nums.length; j++) {
            if(nums[i] === nums[j]) {
                count++;
            }
        }
    }

    return count;
}

function numberOfGoodPairsHM(nums) {
    let count = 0;
    let pairs = {};

    for(let i = 0; i < nums.length; i++) {
        if(pairs[nums[i]] != undefined) {
            count += pairs[nums[i]];
            pairs[nums[i]]++;
        } else {
            pairs[nums[i]] = 1;
        }
    }

    return count;
}

console.log(`${"_".repeat(15)}`)
console.log("Hashmap")
console.log(`${"_".repeat(15)}`)
console.log(numberOfGoodPairsHM([1,2,3,1,1,3]))
console.log(numberOfGoodPairsHM([1,1,1,1]))
console.log(numberOfGoodPairsHM([1,2,3]))

File: 1512_good_pairs/good_pairs.py
# Py3: https://leetcode.com/problems/number-of-good-pairs/submissions/1177989408/
# Solution: https://www.youtube.com/watch?v=BqhDFUo1rjs
from typing import List

def good_pairs(nums: List[int]) -> int:
    count = 0
    for i in range(len(nums)):
        for j in range(len(nums)):
            j = j + i + 1
            if j < len(nums):
                if nums[i] == nums[j]:
                    count += 1
    
    return count

def good_pairs_hm(nums: List[int]) -> int:
    count = 0
    pairs = {}

    for n in nums:
        if n in pairs:
            count += pairs[n]
            pairs[n] += 1
        else:
            pairs[n] = 1
    
    return count

print("_"*15)
print("For loops")
print("_"*15)
print(good_pairs([1,2,3,1,1,3]))
print(good_pairs([1,1,1,1]))
print(good_pairs([1,2,3]))

print("_"*15)
print("Hashmap")
print("_"*15)
print(good_pairs_hm([1,2,3,1,1,3]))
print(good_pairs_hm([1,1,1,1]))
print(good_pairs_hm([1,2,3]))